components-rs/telemetry.rs

```diff
@@ -261,60 +259,66 @@ pub extern "C" fn ddog_sidecar_telemetry_cache_new() -> Box<ShmCacheMap> {
-    cache: &ShmCacheMap,
+pub unsafe extern "C" fn ddog_sidecar_telemetry_config_sent(
+    cache: &mut ShmCacheMap,
     service: CharSlice,
     env: CharSlice,
 ) -> bool {
-    let service_str = service.to_utf8_lossy();
-    let env_str = env.to_utf8_lossy();
-    cache
-        .get(&(service_str.as_ref(), env_str.as_ref()))
-        .map_or(false, |entry| entry.config_sent)
+    ddog_sidecar_telemetry_cache_get_or_update(cache, service, env).config_sent
 }
 
 unsafe fn ddog_sidecar_telemetry_cache_get_or_update<'a>(
     cache: &'a mut ShmCacheMap,
     service: CharSlice,
     env: CharSlice,
-) -> Option<&'a ShmCache> {
-    let cache = &mut *cache;
-
-    let service_str = service.to_utf8_lossy();
-    let env_str = env.to_utf8_lossy();
-
-    let needs_refresh = cache.get(&(service_str.as_ref(), env_str.as_ref())).map_or(true, |entry| {
-        entry.last_updated.elapsed() > Duration::from_secs(1800)
-    });
-
-    if needs_refresh {
-        let shm_path = path_for_telemetry(&service_str, &env_str);
-        if let Ok(mapped) = open_named_shm(&shm_path) {
-            let mut reader = OneWayShmReader::<NamedShmHandle, _>::new(Some(mapped), shm_path);
-            let (_, buf) = reader.read();
+) -> &'a ShmCache {
+    fn refresh_cache(cache: &mut ShmCache) {
+        let (changed, mut buf) = cache.reader.read();
+        if changed {
+            // Cache was reset
+            if buf.is_empty() {
+                cache.reader.clear_reader();
+                let (changed, newbuf) = cache.reader.read();
+                if changed {
+                    buf = newbuf;
+                } else {
+                    cache.config_sent = false;
+                    cache.integrations.clear();
+                    cache.composer_paths.clear();
+                    return;
+                }
+            }
 
             if let Ok((config_sent, integrations, composer_paths)) =
                 bincode::deserialize::<(bool, HashSet<String>, HashSet<PathBuf>)>(buf)
             {
-                let entry = ShmCache {
-                    config_sent,
-                    integrations,
-                    composer_paths,
-                    last_updated: Instant::now(),
-                    reader,
-                };
-                return Some(
-                    cache
-                        .entry(ShmCacheKey(service_str.into(), env_str.into()))
-                        .or_insert(entry),
-                );
+                cache.config_sent = config_sent;
+                cache.integrations = integrations;
+                cache.composer_paths = composer_paths;
             }
         }
+    }
+
+    let service_str = service.to_utf8_lossy();
+    let env_str = env.to_utf8_lossy();
 
-        return None;
+    // I hate you, borrow checker, you get an unsafe from me!
+    if let Some(cached_entry) = (&mut *(cache as *mut ShmCacheMap)).get_mut(&(service_str.as_ref(), env_str.as_ref())) {
```

Comment:
@morrisonlevi If you know how to make that safe, I'm happy (without materializing the key in the case the entry already exists, that is). I absolutely do not get why the borrow checker tells me that I'd take a mut ref twice to the same value here. This if ... ought to be scoping the &mut ref, and allow me to access cache.entry() afterwards.